sdk/identity/azure-identity/src/token_credential_impl.cpp

```diff
@@ -141,10 +146,127 @@ AccessToken TokenCredentialImpl::GetToken(
 }
 
 namespace {
-[[noreturn]] void ThrowJsonPropertyError(std::string const& propertyName)
+std::string const ParseTokenLogPrefix = "TokenCredentialImpl::ParseToken(): ";
+
+std::string TokenAsDiagnosticString(
+    json const& jsonObject,
+    std::string const& accessTokenPropertyName,
+    std::string const& expiresInPropertyName,
+    std::string const& expiresOnPropertyName)
 {
+  std::stringstream ss;
+  ss << "Token JSON";
+
+  if (!jsonObject.is_object())
+  {
+    ss << " is not an object (value='" << jsonObject.dump() << "')";
+  }
+  else
+  {
+    ss << ": Access token property ('" << accessTokenPropertyName << "') ";
+    if (!jsonObject.contains(accessTokenPropertyName))
+    {
+      ss << "is NOT present";
+    }
+    else
+    {
+      auto const& accessTokenProperty = jsonObject[accessTokenPropertyName];
+      if (!accessTokenProperty.is_string())
+      {
+        ss << "is NOT a string (value='" << accessTokenProperty.dump() << "')";
+      }
+      else
+      {
+        ss << "is string (length=" << accessTokenProperty.get<std::string>().length() << ")";
+      }
+    }
+
+    for (auto const& p : {
+             std::pair<char const*, std::string const*>{"relative", &expiresInPropertyName},
+             std::pair<char const*, std::string const*>{"absolute", &expiresOnPropertyName},
+         })
+    {
+      ss << ", " << p.first << " expiration property ('" << *p.second << "') ";
+      if (!jsonObject.contains(*p.second))
+      {
+        ss << "is NOT present";
+      }
+      else
+      {
+        ss << "is present (value='" << jsonObject[*p.second].dump() << "')";
+      }
+    }
+
+    std::map<std::string, json> otherProperties;
+    for (auto const& property : jsonObject.items())
+    {
+      if (property.key() != accessTokenPropertyName && property.key() != expiresInPropertyName
+          && property.key() != expiresOnPropertyName)
+      {
+        otherProperties[property.key()] = property.value();
+      }
+    }
```

Comment:
Is there risk this prints sensitive information?  For example we don't want to print the access token or any other security items

Reply:
We don't print values that are longer than 100 characters, which makes it less probable (tokens are 1KB+).
In any case, the information being sent to logger does live within the same process on the same machine as the actual token, and it is not even included in the exception (suppose you like to print exceptions unfiltered as a reason phrase of a 404 response of your web service).

Reply:
Please see the new commit (unit tests have all sorts of example outputs).